include/xtensor/xtiny.hpp

```diff
@@ -0,0 +1,1999 @@
+        using pointer = value_type *;
+        using const_pointer = const_value_type *;
+        using iterator = value_type *;
+        using const_iterator = const_value_type *;
+        using reverse_iterator = std::reverse_iterator<iterator>;
+        using const_reverse_iterator = std::reverse_iterator<const_iterator>;
+        using size_type = std::size_t;
+        using difference_type = std::ptrdiff_t;
+        using index_type = std::array<size_type, 1>;
+
+        static const bool owns_memory = true;
+        static const bool has_fixed_size = false;
+        static const index_t static_size = runtime_size;
+        static const index_t buffer_size = BUFFER_SIZE;
+
+        template <class NEW_VALUETYPE>
+        using rebind = xtiny<NEW_VALUETYPE, runtime_size, NEW_VALUETYPE[BUFFER_SIZE]>;
+
+        template <index_t NEW_SIZE>
+        using rebind_size = xtiny<value_type, NEW_SIZE < runtime_size ? runtime_size : NEW_SIZE>;
+
+        xtiny_impl();
+        ~xtiny_impl();
+
+        explicit xtiny_impl(size_type n);
+        xtiny_impl(size_type n, const value_type& v);
+        xtiny_impl(size_type n, tags::skip_initialization_tag);
+
+        template <class IT,
+                  class = detail::require_input_iter<IT>>
+        xtiny_impl(IT begin, IT end);
+
+        template <class T>
+        explicit xtiny_impl(std::initializer_list<T> const & v);
+
+        xtiny_impl(xtiny_impl const & v);
+        xtiny_impl(xtiny_impl && v);
+
+        xtiny_impl & operator=(xtiny_impl const & v);
+        xtiny_impl & operator=(xtiny_impl && v);
+
+        void assign(size_type n, const value_type& v);
+
+        template <class IT,
+                  class = detail::require_input_iter<IT>>
+        void assign(IT other_begin, IT other_end);
+
+        reference operator[](size_type i);
+        constexpr const_reference operator[](size_type i) const;
+
+        pointer data();
+        constexpr const_pointer data() const;
+
+        void resize(size_type n);
+
+        size_type capacity() const;
+        size_type size() const;
+        size_type max_size() const;
+        index_type shape() const;
+        bool on_stack() const;
+
+        iterator begin();
+        const_iterator begin() const;
+
+        reverse_iterator rbegin();
+        const_reverse_iterator rbegin() const;
+
+        void swap(xtiny_impl & other);
+
+      protected:
+        static const bool may_use_uninitialized_memory = xtrivially_default_constructible<value_type>::value;
+
+        /* allocate() assumes that m_size is already set,
+           but no memory has been allocated yet */
+        void allocate(value_type const & v = value_type());
+        void allocate(tags::skip_initialization_tag);
+        template <class IT,
+                  class = detail::require_input_iter<IT>>
+        void allocate(IT other_begin);
+
+        void deallocate();
+
+        allocator_type m_allocator;
+        size_type m_size;
```

Comment:
did you have the chance to benchmark this against storing begin/end pointers? At least llvm's small vector stores begin and end pointers, and it might be faster in certain cases.

Reply:
I didn't benchmark this specific code. But I've consistently found that loops over integers are as fast (or even faster) as loops over iterators. For example, one would think that
```c++
auto iter = v.begin();
for(size_t k=0, k < v.size(); ++k, ++iter)
```
is slower than
```c++
for(auto iter = v.begin(); iter != v.end(); ++iter)
```
because one increments two variables. But actually the two versions are equally fast, or the first is faster because compilers seem to be better at optimizing integer loops. However, please confirm these statements by independent measurements.